tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/OptimizerCore.scala

```diff
@@ -2183,6 +2243,32 @@ private[optimizer] abstract class OptimizerCore(
           case _               => default
         }
 
+      case String_+ =>
+        val lhs1 = foldToStringForString_+(lhs)
+        val rhs1 = foldToStringForString_+(rhs)
+        @inline def stringDefault = BinaryOp(String_+, lhs1, rhs1)
+        (lhs1, rhs1) match {
+          case (StringLiteral(s1), StringLiteral(s2)) => 
+            StringLiteral(s1 + s2)
+          case (l, r @ StringLiteral("")) =>
+            foldBinaryOp(op, r, l)
```

Comment:
Directly use `lhs1` and `rhs1` instead of `l` and `r` (and then you don't need to capture them in the pattern). Same in the other cases.